### PR TITLE
Reference: Document block dependencies property

### PR DESCRIPTION
**Pull request by bmarkons**
This PR introduces `dependencies in blocks` section within pipeline YAML reference.

It explains how the property is used along with one example snippet.

Let me know if there is something you would like to add within a reference.

### CODE REVIEW

File: pipeline-yaml_5b5744d40428631d7a8940a9.md [hunk 1]
```diff
@@ -371,10 +372,58 @@ Error: "There are at least two blocks with same name: Build Go project"
...
+  machine:
+    type: e1-standard-2
+    os_image: ubuntu1804
+
+blocks:
+  - name: "A"
+    dependencies: []
+    task:
+      jobs:
+      - name: "A"
```
Comment (DamjanBecirovic) [inline pipeline-yaml_5b5744d40428631d7a8940a9.md]: Maybe these names should be changed to `Job A` etc. so people would not be confused about which name they should put in `dependencies` list .

Comment (bmarkons) [reply pipeline-yaml_5b5744d40428631d7a8940a9.md]: This is very good notice. I will change it.

Also I consciously decided to name Semaphore entities in a highly generic way, in order not to confuse people with some specific use case who might not be familiar with it. Feedback about this decision is also welcome.

Thanks!

**Comment by darkofabijan** [comment]
@markoa please take a look.

File: pipeline-yaml_5b5744d40428631d7a8940a9.md [hunk 2]
```diff
@@ -77,12 +78,12 @@ version: v1.0
 ## name in preface
 
 The `name` property is a Unicode string that assigns a name to a Semaphore
-2.0 project and is optional. However, you should always give descriptive names
-to your Semaphore 2.0 projects.
+2.0 pipeline and is optional. However, you should always give descriptive names
```
Comment (markoa) [inline pipeline-yaml_5b5744d40428631d7a8940a9.md]: Let's use the opportunity to remove "2.0" from this paragraph.

File: pipeline-yaml_5b5744d40428631d7a8940a9.md [hunk 3]
```diff
@@ -371,10 +372,59 @@ Error: "There are at least two blocks with same name: Build Go project"
 Semaphore assigns its own unique names to nameless `blocks` items, which are
 displayed in the Semaphore 2.0 UI.
 
+### dependencies in blocks
+
+The `dependencies` property of a block is used to define the flow of execution
+between blocks within a pipeline.
```
Comment (markoa) [inline pipeline-yaml_5b5744d40428631d7a8940a9.md]: Prefer active over passive voice. This applies to all writing.

Comment (markoa) [reply pipeline-yaml_5b5744d40428631d7a8940a9.md]: It seems like I might want to use this when running blocks in parallel, but it's not clear. Can you rephrase?

Assume that the reader is trying to figure out if this is relevant to them from the first paragraph.

File: pipeline-yaml_5b5744d40428631d7a8940a9.md [hunk 4]
```diff
@@ -371,10 +372,59 @@ Error: "There are at least two blocks with same name: Build Go project"
...
+
+To specify that block with name `Block C` needs to be run once the
+blocks with names `Block A` and `Block B` finish, the `dependencies` property of
+the block with name `Block C` needs to be populated with corresponding names
+of the blocks which block with name `Block C` depends on.
+In the following example blocks with names `Block A` and `Block B` would be executed in parallel at the
+very beginning of a pipeline.
```
Comment (markoa) [inline pipeline-yaml_5b5744d40428631d7a8940a9.md]: This is very hard to read, first sentence is in 4 lines.

File: pipeline-yaml_5b5744d40428631d7a8940a9.md (hunk 3 above)
Comment (markoa) [reply pipeline-yaml_5b5744d40428631d7a8940a9.md]: There's almost always a way to say the same thing in active voice, which sounds more friendly, without losing the technical rigor.

For example:

_When your pipeline is running blocks in parallel, you can use the `dependencies` property to define the flow of execution for subsequent blocks._

File: pipeline-yaml_5b5744d40428631d7a8940a9.md (hunk 4 above)
Comment (markoa) [reply pipeline-yaml_5b5744d40428631d7a8940a9.md]: Please rewrite with less words, taking into account my earlier comment about active voice.

File: pipeline-yaml_5b5744d40428631d7a8940a9.md [hunk 5]
```diff
@@ -371,10 +372,59 @@ Error: "There are at least two blocks with same name: Build Go project"
...
+    task:
+      jobs:
+      - name: "Job C"
+        commands:
+          - echo "output"
+```
+
+*Note*: When `dependencies` property is used to define the names of the blocks which certain block
+depends on, property becomes mandatory for all other blocks within a pipeline
+definition. Otherwise the YAML validation error is being thrown.
```
Comment (markoa) [inline pipeline-yaml_5b5744d40428631d7a8940a9.md]: This is hard to understand. Can you provide an example of what wouldn't work?

File: pipeline-yaml_5b5744d40428631d7a8940a9.md [hunk 6]
```diff
@@ -371,10 +372,59 @@ Error: "There are at least two blocks with same name: Build Go project"
...
+In case `dependencies` property is not used in any of the blocks within a
```
Comment (markoa) [inline pipeline-yaml_5b5744d40428631d7a8940a9.md]: This is something else, so make it a separate paragraph.

File: pipeline-yaml_5b5744d40428631d7a8940a9.md (hunk 5 above)
Comment (markoa) [reply pipeline-yaml_5b5744d40428631d7a8940a9.md]: I think it means that if you write `dependencies` in one block, then you have to write it in each block?

File: pipeline-yaml_5b5744d40428631d7a8940a9.md [hunk 7]
```diff
@@ -371,10 +372,59 @@ Error: "There are at least two blocks with same name: Build Go project"
...
+### dependencies in blocks
+
+The `dependencies` property of a block is used to define the flow of execution
+between blocks within a pipeline.
+
+To specify that block with name `Block C` needs to be run once the
+blocks with names `Block A` and `Block B` finish, the `dependencies` property of
+the block with name `Block C` needs to be populated with corresponding names
+of the blocks which block with name `Block C` depends on.
+In the following example blocks with names `Block A` and `Block B` would be executed in parallel at the
```
Comment (markoa) [inline pipeline-yaml_5b5744d40428631d7a8940a9.md]: Can you think of an additional example? I worry that people may not understand the full power of the feature from this one example.

File: pipeline-yaml_5b5744d40428631d7a8940a9.md (hunk 5 above)
Comment (bmarkons) [reply pipeline-yaml_5b5744d40428631d7a8940a9.md]: Yes. I reduced the number of words and added one example when the specification is invalid.

File: pipeline-yaml_5b5744d40428631d7a8940a9.md (hunk 7 above)
Comment (bmarkons) [reply pipeline-yaml_5b5744d40428631d7a8940a9.md]: The whole reference section of this property aims to explain simple mechanism upon one can build even the most complex workflows. In the last sentence, the reader is linked to the [repo with workflows examples](https://github.com/semaphoreci-demos/semaphore-demo-workflows) where the most common are specified.

**Comment by bmarkons** [comment]
@markoa Let me know if you think something more should appear in this section.
